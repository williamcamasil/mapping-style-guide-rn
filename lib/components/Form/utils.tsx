import type { FormApi, MutableState } from 'final-form';

export const hasValue = (value: any) => {
  if (value === undefined) return false;
  if (value === null) return false;

  if (Array.isArray(value)) return value.length > 0;

  if (value instanceof Date) return true;

  switch (typeof value) {
    case 'object':
      return Object.keys(value).length > 0;
    case 'number':
    case 'boolean':
      return true;
    case 'string':
    default:
      return Boolean(value);
  }
};

export function getFieldErrorState(
  invalid: boolean | undefined,
  touched: boolean | undefined,
  error?: any,
  submitError?: any,
  helpMessage?: string,
) {
  const fieldError = error || submitError;

  const showError = Boolean(invalid && touched);
  const message = showError && typeof fieldError === 'string'
    ? fieldError
    : helpMessage;

  return {
    showError,
    message,
  };
}

type SetTouchedArguments = Parameters<(name: string, touched: boolean) => void>;

export function setFormFieldTouched<T>(form: FormApi<T>, fieldName: string, touched: boolean) {
  /* istanbul ignore next */
  if (!form.mutators.setFieldTouched) {
    console.warn('O form n√£o possui o mutator setFieldTouched');
    return;
  }
  form.mutators.setFieldTouched(fieldName, touched);
}

export function setFieldTouched<FormValues, InitialFormValues = Partial<FormValues>>(
  args: SetTouchedArguments,
  state: MutableState<FormValues, InitialFormValues>,
) {
  const [name, touched] = args;
  const field = state.fields[name];
  if (field) {
    field.touched = Boolean(touched);
  }
}

