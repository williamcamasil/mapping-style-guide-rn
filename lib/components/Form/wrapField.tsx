import React, { useCallback } from 'react';

import { FieldState } from 'final-form';
import { Field, FieldMetaState, FieldRenderProps } from 'react-final-form';

import { hasValue } from './utils';

export type WrapFieldProps<FieldValue> = FieldMetaState<FieldValue> & {
  value?: FieldValue | null;
  onChange?: (value: FieldValue | null) => void;
  onBlur?: (event?: any) => void;
  onFocus?: (event?: any) => void;
  required?: boolean;
};

export type FieldValidator<FieldValue, FormValues = any, InitialFormValues = Partial<FormValues>> = (
  value: FieldValue,
  allValues?: InitialFormValues,
  meta?: FieldState<FieldValue>
) => any | Promise<any>;

export type WrappedFieldRenderProps<FieldValue> = {
  /**
   * Nome do field que será retornado no objeto `values` ao submeter o formulário.
   */
  name: string;
  /**
   * Função que valida o estado do field.
   * Se o field está válido, então deve retornar `undefined`.
   * Caso contrário, deve retornar a `string` com a mensagem de erro.
   */
  validate?: FieldValidator<FieldValue>;
  /**
   * Função que faz o parse do valor informado antes de salvar no `state` do
   * formulário:
   *
   * field -> parse -> formState
   */
  parse?: (value: FieldValue, name: string) => FieldValue;
  /**
   * Função que formata o valor do campo para renderizar na tela:
   *
   * formState -> format -> field
   */
  format?: (value: FieldValue, name: string) => FieldValue;
  /**
   * Quando requerido, o `validate` é aplicado mesmo com o field vazio.
   * Se não for requerido, então o `validate` é aplicado apenas quando o field tiver um valor.
   */
  required?: boolean;
  /**
   * valor inicial de um campo
  */
  initialValue?: FieldValue;
};

export default function wrapField<
  Props,
  FieldValue = string | undefined,
  ExternalProps = Omit<Props, keyof WrapFieldProps<FieldValue>>,
>(renderField: (componentProps: ExternalProps, fieldProps: WrapFieldProps<FieldValue>) => React.ReactElement) {

  type FieldWrapperPropsType = WrappedFieldRenderProps<FieldValue> & ExternalProps;

  function FieldWrapper({
    name,
    validate,
    parse,
    format,
    required,
    initialValue,
    ...othersFieldProps
  }: FieldWrapperPropsType) {
    const renderFieldProps = useCallback((renderProps: FieldRenderProps<FieldValue>) => renderField(
      /**
       * Foi necessário desativar passar any aqui devido a uma limitação de tipagem do FinalForm.
       * Assim é possível inferir o tipo corretamente ao componente retornado pelo "wrapField".
       */
      (othersFieldProps as any),
      {
        ...renderProps.input,
        ...renderProps.meta,
        required,
      },
    ), [othersFieldProps, required]);

    const handleValidation: FieldValidator<FieldValue> = useCallback((value, values, meta) => {
      if (validate && (required || hasValue(value))) {
        return validate(value, values, meta);
      }
      return undefined;
    }, [validate, required]);

    return (
      <Field<FieldValue>
        name={name}
        render={renderFieldProps}
        validate={handleValidation}
        parse={parse}
        format={format}
        initialValue={initialValue}
      />
    );
  }

  return FieldWrapper;
}
