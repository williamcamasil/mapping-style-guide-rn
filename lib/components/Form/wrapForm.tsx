import React, {
  ComponentType, useCallback, useRef,
} from 'react';

import {
  FormApi, Mutator, SubmissionErrors, ValidationErrors,
} from 'final-form';
import { delay } from 'mapping-context-rn';
import { Form, FormRenderProps } from 'react-final-form';

import { setFieldTouched } from './utils';

export type OnSubmitFormType<FormValues, InitialFormValues = Partial<FormValues>> = (
  values: FormValues,
  form: FormApi<FormValues, InitialFormValues>,
) => SubmissionErrors | Promise<SubmissionErrors | void> | void;

type WrapFormRenderProps<FormValues, InitialFormValues> = Omit<FormRenderProps<FormValues, InitialFormValues>, 'handleSubmit'> & {
  handleSubmit: (
    submitCallback: OnSubmitFormType<FormValues, InitialFormValues>
  ) => () => void;
};

type WrapFormOptionsType<FormValues, InitialFormValues> = {
  form?: FormApi<FormValues, InitialFormValues>;
  initialValues?: InitialFormValues;
  validate?: (
    values: FormValues,
  ) => ValidationErrors | Promise<ValidationErrors>;
  validateOnBlur?: boolean;
  mutators?: { [key: string]: Mutator<FormValues, InitialFormValues> };
};

export default function wrapForm<
  Props,
  FormValues,
  InitialFormValues = Partial<FormValues>,
>(
  FormComponent: ComponentType<Props & WrapFormRenderProps<FormValues, InitialFormValues>>,
  options: WrapFormOptionsType<FormValues, InitialFormValues> = {},
) {

  type FormWrapperPropsType = Props & {
    form?: FormApi<FormValues, InitialFormValues>;
  };

  const defaultOptions: WrapFormOptionsType<FormValues, InitialFormValues> = {
    ...options,
    mutators: {
      setFieldTouched,
      ...options.mutators,
    },
  };

  function FormWrapper({ form: injectedForm, ...formWrapperProps }: FormWrapperPropsType) {
    const handleSubmitRef = useRef<FormRenderProps<FormValues, InitialFormValues>['handleSubmit']>();
    const submitCallbackRef = useRef<OnSubmitFormType<FormValues, InitialFormValues>>();

    const handleSubmitCreator = useCallback((submitCallback: OnSubmitFormType<FormValues, InitialFormValues>) => {
      submitCallbackRef.current = submitCallback;

      return () => handleSubmitRef.current?.();
    }, []);

    const handleFormSubmit = useCallback(async (
      values: FormValues,
      form: FormApi<FormValues, InitialFormValues>,
    ) => {
      // aguarda 1ms para que o form possa atualizar seu estado interno antes de submeter
      await delay(1);
      return submitCallbackRef.current?.(values, form);
    }, []);

    const renderForm = useCallback(({
      handleSubmit,
      ...othersRenderProps
    }: FormRenderProps<FormValues, InitialFormValues>) => {
      /**
       * Armazena a função de submit que irá disparar a submissão do formulário.
       */
      handleSubmitRef.current = handleSubmit;

      return (
        <FormComponent
          {...formWrapperProps as Props}
          {...othersRenderProps}
          /**
           * Passa para o componente a função que irá "criar" a função de submissão.
           */
          handleSubmit={handleSubmitCreator}
        />
      );
    }, [formWrapperProps, handleSubmitCreator]);

    return (
      <Form<FormValues, InitialFormValues>
        {...defaultOptions}
        onSubmit={handleFormSubmit}
        form={injectedForm ?? defaultOptions?.form}
      >
        {renderForm}
      </Form>
    );
  }

  FormWrapper.displayName = `FormWrapper(${FormComponent.displayName})`;

  return FormWrapper;
}
