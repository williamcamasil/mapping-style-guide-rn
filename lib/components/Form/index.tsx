import React from 'react';

import { FormApi, Mutator, ValidationErrors } from 'final-form';
import { Form as FinalForm, FormRenderProps } from 'react-final-form';

import { OnSubmitFormType } from './wrapForm';

export { default as wrapForm } from './wrapForm';
export { default as wrapField } from './wrapField';
export { hasValue, getFieldErrorState, setFormFieldTouched } from './utils';

export type { OnSubmitFormType } from './wrapForm';
export type {
  WrappedFieldRenderProps, WrapFieldProps, FieldValidator,
} from './wrapField';

export type FormPropsType<FormValues, InitialFormValues> = {
  children: (props: FormRenderProps<FormValues, InitialFormValues>) => React.ReactNode;
  onSubmit: OnSubmitFormType<FormValues, InitialFormValues>;
  form?: FormApi<FormValues, InitialFormValues>;
  initialValues?: InitialFormValues;
  validate?: (
    values: FormValues,
  ) => ValidationErrors | Promise<ValidationErrors>;
  validateOnBlur?: boolean;
  mutators?: { [key: string]: Mutator<FormValues, InitialFormValues> };
};

function Form<FormValues, InitialFormValues = Partial<FormValues>>(
  props: FormPropsType<FormValues, InitialFormValues>,
): JSX.Element {
  return (
    <FinalForm<FormValues, InitialFormValues> {...props} />
  );
}

export default Form;
