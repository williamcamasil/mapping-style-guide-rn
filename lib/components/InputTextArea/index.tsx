import React from 'react';
import {
  StyleProp,
  View,
  ViewStyle,
} from 'react-native';

import { useViewStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { getFieldErrorState, wrapField } from '../Form';
import InputText, { InputTextPropsType } from '../InputText';

export type InputTextAreaPropsType = InputTextPropsType & {
  testID?: string;
  containerTextAreaStyle?: StyleProp<ViewStyle>;
  textAreaStyle?: StyleProp<ViewStyle>;
  numberOfLines?: number;
};

const InputTextArea = ({
  testID,
  containerTextAreaStyle,
  textAreaStyle,
  numberOfLines,
  ...others
}: InputTextAreaPropsType) => {

  const inputTextAreaStyle = useViewStyles(() => [
    {
      textAlignVertical: 'top',
      justifyContent: 'flex-start',
      height: 'auto',
    },
    textAreaStyle,
  ], [textAreaStyle]);

  return (
    <View style={containerTextAreaStyle}>
      <InputText
        {...others}
        multiline
        textArea
        testID={testID}
        numberOfLines={numberOfLines}
        style={inputTextAreaStyle}
      />
    </View>
  );
};

InputTextArea.defaultProps = {
  numberOfLines: 2,
};

InputTextArea.Field = withTheme(wrapField<InputTextAreaPropsType>((props, fieldProps) => {
  const {
    invalid, touched, error, submitError, value, onChange, onBlur, onFocus,
  } = fieldProps;

  const { showError, message } = getFieldErrorState(
    invalid,
    touched,
    error,
    submitError,
    props.help,
  );

  return (
    <InputTextArea
      {...props}
      value={value || ''}
      onChange={onChange}
      onBlur={onBlur}
      onFocus={onFocus}
      error={showError}
      help={message}
    />
  );
}));

export default withTheme(InputTextArea);
