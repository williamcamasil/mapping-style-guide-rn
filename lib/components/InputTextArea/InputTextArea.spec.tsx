import React from 'react';

import { render, fireEvent } from '@testing-library/react-native';
import { createForm } from 'final-form';

import InputTextArea from '.';
import Form from '../Form';

describe('inputTextArea snapshot', () => {
  it('should match default snapshot', () => {
    const tree = render(
      <InputTextArea
        label="Text area label"
        value=""
      />,
    )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match error snapshot', () => {
    const tree = render(
      <InputTextArea
        label="Text area label"
        value=""
        error
        help="Error message"
      />,
    )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('tests with inputTextArea Field', () => {
  it('should valid inputTextArea', async () => {
    const form = createForm({
      onSubmit: jest.fn(),
    });

    const component = render(
      <Form
        onSubmit={jest.fn()}
        form={form}
      >
        {() => (
          <InputTextArea.Field
            testID="text-area-input"
            name="InputTextArea"
            label="Text area"
            required
            validate={value => (value?.length === 6 ? undefined : 'Erro')}
          />
        )}
      </Form>,
    );

    expect(form.getState().valid).toBeFalsy();

    const inputTextArea = component.getByTestId('text-area-input');
    fireEvent(inputTextArea, 'changeText', '123456');

    expect(form.getState().valid).toBeTruthy();
  });
});
