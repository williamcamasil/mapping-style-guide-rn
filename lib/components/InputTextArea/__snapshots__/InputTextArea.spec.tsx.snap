// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inputTextArea snapshot should match default snapshot 1`] = `
<View>
  <View>
    <TextInput
      allowFontScaling={false}
      editable={true}
      multiline={true}
      numberOfLines={2}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="Text area label"
      placeholderTextColor="#A4A7AF"
      secureTextEntry={false}
      style={
        {
          "backgroundColor": "#FFFFFF",
          "color": "#676D79",
          "fontFamily": "Averta CY",
          "fontSize": 16,
          "fontWeight": "normal",
          "height": "auto",
          "justifyContent": "flex-start",
          "letterSpacing": -0.02,
          "lineHeight": 16,
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingTop": 4,
          "textAlignVertical": "top",
        }
      }
      value=""
    />
  </View>
</View>
`;

exports[`inputTextArea snapshot should match error snapshot 1`] = `
<View>
  <View>
    <TextInput
      allowFontScaling={false}
      editable={true}
      multiline={true}
      numberOfLines={2}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="Text area label"
      placeholderTextColor="#A4A7AF"
      secureTextEntry={false}
      style={
        {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#E53930",
          "color": "#676D79",
          "fontFamily": "Averta CY",
          "fontSize": 16,
          "fontWeight": "normal",
          "height": "auto",
          "justifyContent": "flex-start",
          "letterSpacing": -0.02,
          "lineHeight": 16,
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingTop": 4,
          "textAlignVertical": "top",
        }
      }
      value=""
    />
    <Text
      allowFontScaling={false}
      style={
        {
          "color": "#E53930",
          "flexShrink": 1,
          "fontFamily": "Averta CY",
          "fontSize": 14,
          "fontWeight": "normal",
          "letterSpacing": -0.02,
          "lineHeight": 14,
          "marginLeft": 0,
          "marginTop": 12,
        }
      }
    >
      Error message
    </Text>
  </View>
</View>
`;
